### PR TITLE
Move Statistics stdlib module to external repository

### PR DESCRIPTION
**Pull request by nalimilan**
The goal is to be able to import features from StatsBase (notably https://github.com/JuliaLang/julia/pull/31395) and keep history without messing too much with the Julia history. And of course, once stdlib modules can be updated, to allow using new features from old Julia versions.

I've tried to follow what is done for Pkg.jl, I hope I got everything right.

See https://github.com/JuliaStats/Statistics.jl/issues/1 for how the files and history were extracted from the Julia repo.

### CODE REVIEW

File: stdlib/Statistics.version
Comment (c42f) [inline stdlib/Statistics.version]: Maybe slightly OT, but I noticed that it also seems possible to use the git tree SHA with the github API rather than the commit which I believe is also the way Pkg identifies the code for a version. f52ff30330bf9c2ddf91bc731c0cad78d9ec5b49 is the tree sha for this commit.

**Review by c42f — COMMENTED** [review]
Good to see that the hacks I made to generalize `STDLIBS_EXT` (quite some time ago now) actually work :-D The technical detail of this looks fine to me.

One query - is there consensus that this stdlib should live in `JuliaStats`? (Sounds reasonable to me; just wondering whether it's been bikeshedded etc.)

File: stdlib/Statistics.version
Comment (fredrikekre) [reply stdlib/Statistics.version]: It works (we tried it) , but the tarballs obtained from GitHub doesn't have a consistent hash.

Comment (c42f) [reply stdlib/Statistics.version]: Hmm, odd. I just tried it and downloading https://api.github.com/repos/JuliaStats/Statistics.jl/tarball/f52ff30330bf9c2ddf91bc731c0cad78d9ec5b49 does appear to untar to a tree which is binary identical to Statistics.jl checked out at da6057b. I assume you refer to something else which is broken?

Comment (fredrikekre) [reply stdlib/Statistics.version]: I mean the hashsum of the tarball itself.

**Comment by nalimilan** [comment]
> One query - is there consensus that this stdlib should live in `JuliaStats`? (Sounds reasonable to me; just wondering whether it's been bikeshedded etc.)

I asked on Slack and nobody commented except @ararslan, so this PR reflects the consensus of 2 people. :-)

File: stdlib/Statistics.version
Comment (nalimilan) [reply stdlib/Statistics.version]: So should I change something or not?

Comment (c42f) [reply stdlib/Statistics.version]: No it looks like Pkg goes by the commit hash anyway.

**Comment by c42f** [comment]
> so this PR reflects the consensus of 2 people

Heh :-)

FWIW I was expecting to throw Logging into JuliaLang/Logging.jl at some stage but it doesn't have an obvious alternative home.

What are the pros and cons of having stdlibs in JuliaLang vs elsewhere? It seems mainly about management of permissions and possibly access to infrastructure? For example, does it make sense for stdlibs to eventually have access to the core benchmarking infrastructure and CI runners? @staticfloat any thoughts?

**Comment by ararslan** [comment]
Where the repository actually lives doesn't really matter; by virtue of it being a stdlib, it's pulled in and bundled with Julia when Julia is built.

For testing, that means that the tests for stdlibs are always run on the buildbot infrastructure, since a build of Julia contains all stdlibs, unless tests are explicitly disabled. (That was done recently for Pkg.)

As for benchmarking, Nanosoldier's code and infrastructure is set up _only_ to run on this repo. It does a full build of Julia so it will end up pulling in stdlibs anyway, which can continue to be benchmarked via BaseBenchmarks, just like the ones that currently live in this repo are.

**Comment by c42f** [comment]
> Where the repository actually lives doesn't really matter; by virtue of it being a stdlib, it's pulled in and bundled with Julia when Julia is built.

That's fine for releases, but not really any good for development in the longer term. The stdlibs need their own CI just like any other package.

**Comment by ararslan** [comment]
Pkg seems to be getting by with Travis and AppVeyor.

**Comment by nalimilan** [comment]
Yeah I don't think where the package lives matters a lot. We can even move it very easily without breaking anything. If Julia developers prefer all stdlib packages to live in JuliaLang, that's also fine with me.

**Comment by ViralBShah** [comment]
I think having the stdlibs be in JuliaLang but in separate reps would be good. I would love to refactor out linalg and sparse similarly.

**Comment by musm** [comment]
I agree with Viral, it would make it easier for us to view related issues if they were all in the same JuliaLang repo (especially for contributors).  It would be even more awesome if  github implemented a merge issue tracker from different repos, so that all stdlib issues could be grouped together.

**Comment by c42f** [comment]
I'll admit I've also got a bias to have the stdlib repos in JuliaLang, but I'm not sure it's founded.

That's why I'm asking whether there's any substantive difference about where these live. I think it ultimately comes down to who has commit access. There's two different needs here:
* Timely and non-disruptive maintenance fixes, as related to releases of the standard Julia *distribution* (Base + stdlibs). The language itself will likely depend even more intimately on some stdlibs in the future. For example, CSTParser or whatever it evolves into. From this perspective it makes sense to put stdlibs in JuliaLang so that core contributors automatically have access. The more detached the stdlibs become, the more they end up like other external dependencies which we currently carry patches for.
* Community-driven development: The community outside the core language contributors is super important in knowing what is needed, how to build it and in driving that development forward. The community needs to be empowered to build what is needed or the stdlibs will stagnate. From this perspective it makes sense to embed stdlibs in the part of the community which knows them best (`Statistics` going into `JuliaStats` being a good example).

With everything being open source and the ability to add external contributors to individual repos there may not be much practical difference here. There might be a symbolic difference though :-)

**Comment by ararslan** [comment]
JuliaStats seems like a natural home for it based on the expected contributors and audience, but I also don't care that much, so I'm fine with whatever.

**Comment by staticfloat** [comment]
I think the tension here is between domain experts (e.g. contributors to `JuliaStatistics`) and core language folks/release managers (e.g. Alex, myself, etc....).  As long as both groups have the ability to do things like tag a release (this is not very important now, but may become more important once we further decouple stdlibs from the Julia build process) I don't think there are any problems with hosting this in a separate org like `JuliaStatistics`.

**Comment by ViralBShah** [comment]
I myself have the same tension. I'd like a bunch of stuff in `JuliaSparse` to have that community maintain these packages. But so long as it is an stdlib, I like having it all in one org since there are other non-sparse matrix maintenance tasks that get carried out by others. Is security a concern? We have a managed group of contributors on JuliaLang, with 2FA enforced and so on - but other orgs might not do the same.

**Review by c42f — APPROVED** [review]
Looks good to me. FWIW I restarted CI on windows though it would seem failures could hardly be related.

**Comment by nalimilan** [comment]
Thanks. I've added checksums which are needed AFAICT.

**Comment by nalimilan** [comment]
Now that's funny. Before the last commit I got this error on [32-bit Linux](https://build.julialang.org/#/builders/35/builds/6699/steps/7/logs/stdio):
```
/buildworker/worker/package_linux32/build/deps/tools/jlchecksum /tmp/srccache/Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz
===============================================================================
  ERROR: sha512 checksum failure on Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz, should be:
      982a1c3e0ee4d6dc5935c1db6d3e6dd67595cd95d867d6a0bf9bb88300b97a2f
      8aa7e91bf4b9461570a0b7a17f0f80006e8af4a1fd65c0529a93fab29a0760be
  But `sha512sum /tmp/srccache/Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz | awk '{ print $1; }'` results in:
      e3c490ef7f49b6d01a26406cac128a35d51ef5c66b470a3cb3bf924662959525
      780cfe1cc463181c4ee4b698c40eaf13e79f62c6a2fe31acb1d30cb6c2dbcc2f
  This can happen due to bad downloads or network proxies, please check your
  network proxy/firewall settings and delete
  /tmp/srccache/Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz
  to force a redownload when you are ready
===============================================================================
Makefile:27: recipe for target 'Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2/source-extracted' failed
```
And now with the last commit I get this error [on 64-bit Linux](https://build.julialang.org/#/builders/53/builds/6764/steps/7/logs/stdio):
```
/buildworker/worker/package_linux64/build/deps/tools/jlchecksum /tmp/srccache/Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz
===============================================================================
  ERROR: sha512 checksum failure on Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz, should be:
      e3c490ef7f49b6d01a26406cac128a35d51ef5c66b470a3cb3bf924662959525
      780cfe1cc463181c4ee4b698c40eaf13e79f62c6a2fe31acb1d30cb6c2dbcc2f
  But `sha512sum /tmp/srccache/Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2.tar.gz | awk '{ print $1; }'` results in:
      982a1c3e0ee4d6dc5935c1db6d3e6dd67595cd95d867d6a0bf9bb88300b97a2f
      8aa7e91bf4b9461570a0b7a17f0f80006e8af4a1fd65c0529a93fab29a0760be
  This can happen due to bad downloads or network proxies, please check your
  network proxy/firewall settings and delete
[ ! \( -e Pkg-6c978c097d15cc868335776ff0b01be22d15de98/ -o -h Pkg-6c978c097d15cc868335776ff0b01be22d15de98/ \) ] || rm -r Pkg-6c978c097d15cc868335776ff0b01be22d15de98/
/buildworker/worker/package_linux64/build/deps/tools/jlchecksum: line 41: realpath: command not found
  
  to force a redownload when you are ready
===============================================================================
make[1]: *** [Statistics-da6057baf849cbc803b952ef7adf979ae3a9f9d2/source-extracted] Error 2
```

**Comment by nalimilan** [comment]
I've changed the PR to use another commit, and CI passed. I really think something weird happened due to moving the project from JuliaStats to JuliaLang, either on the GitHub side or on the cache side.

**Comment by nalimilan** [comment]
I guess it looks good now? Not sure what's going on with the Windows tests (failing download).

**Comment by c42f** [comment]
Yeah those failures are pretty weird: basically a timeout when downloading julia-installer.exe. It seems to download the whole thing or almost the whole thing, then gets stuck. (The last log message from `curl` is at 57.2M and the whole thing is 58M; I can download the x64 version locally to a linux machine with the same `curl` command.)

Anyway, the windows builds passed which means the changes here should be fine given they apply to the build process. A failed download during testing on one platform doesn't seem worrying.

**Comment by c42f** [comment]
I restarted the win64 tester and it seems stuck again at the same download. @staticfloat any thoughts?

**Comment by staticfloat** [comment]
Okay I debugged this a bit today, and I think the issue here is that this PR just _barely_ managed to dodge the commit on Sept. 28th that changed the installer to use Inno instead of NSIS.  Can someone rebase this off of the latest `master` to see if that will play nicely with the new buildbots?

**Comment by nalimilan** [comment]
That fixed the failures! I'll merge in a few days if nobody objects.

**Comment by ViralBShah** [comment]
Should we transfer the statistics related issues to that repo?

**Comment by KristofferC** [comment]
Imo yes.

**Comment by ViralBShah** [comment]
I'm looking forward to doing the same then with sparse matrix packages and eventually linear alebra.

**Comment by nalimilan** [comment]
Yes, definitely.

**Pull request merged**
